contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -3113,6 +3118,359 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+	char	*max_identity_value_str = NULL;
+	char	*query = NULL;
+	char	*attname;
+	char	*token;
+	const char	*schema_name;
+	const char	*nsp_name;
+	const char	*user;
+	const char	*guest_role_name;
+	const char	*dbo_role_name;
+	const char	*login;
+	int64	max_identity_value = 0;
+	int64	cur_identity_value = 0;
+	int	attnum;
+	int	rc = 0;
+	int64	reseed_value = 0;
+	Oid	nsp_oid;
+	Oid 	table_oid;
+	Oid	seqid = InvalidOid;
+	Oid	current_user_id = GetUserId();
+	volatile bool cur_value_is_null = true;
+	bool	login_is_db_owner;
+	char	message[200];
+	bool	is_float_value;
+	bool    is_cross_db = false;
+
+
+	if(dbcc_stmt.new_reseed_value)
+	{
+		/* If float value is passed as reseed_value, only part before decimal is considered */
+		is_float_value = strchr(dbcc_stmt.new_reseed_value, '.') != NULL;
+
+		if (is_float_value)
+		{
+			if (dbcc_stmt.new_reseed_value[0] == '.' || 
+				(dbcc_stmt.new_reseed_value[0] == '-' && dbcc_stmt.new_reseed_value[1] == '.'))
+				reseed_value = 0;
+			else
+			{
+				token = strtok(dbcc_stmt.new_reseed_value, ".");
+				reseed_value = pg_strtoint64(token);
+				pfree(token);
+			}
+		}
+		else
+			reseed_value = pg_strtoint64(dbcc_stmt.new_reseed_value);
+	}
+	
+	db_name = get_cur_db_name();
+	if (dbcc_stmt.db_name)
+	{
+		if (!DbidIsValid(get_db_id(dbcc_stmt.db_name)))
+		{
+			ereport(ERROR,
+			(errcode(ERRCODE_UNDEFINED_DATABASE),
+				errmsg("database \"%s\" does not exist", dbcc_stmt.db_name)));
+		}
+		if (pg_strncasecmp(db_name, dbcc_stmt.db_name, NAMEDATALEN) != 0)
+		{
+			is_cross_db = true;
+			pfree(db_name);
+			db_name = pstrdup(dbcc_stmt.db_name);
+		}
+	}
+
+	user = get_user_for_database(db_name);
+	login_is_db_owner = 0 == strncmp(GetUserNameFromId(GetSessionUserId(), false),
+										get_owner_of_db(db_name), NAMEDATALEN);
+
+	/* Raise an error if the login does not have access to the database */
+	if(is_cross_db)
+	{
+		if (user)
+			SetCurrentRoleId(GetSessionUserId(), false);
+		else
+		{
+			login = GetUserNameFromId(GetSessionUserId(), false);
+			pfree(db_name);
+			ereport(ERROR,
+                    	(errcode(ERRCODE_UNDEFINED_DATABASE),
+                    		errmsg("The server principal \"%s\" is not able to access "
+                            	"the database \"%s\" under the current security context",
+                           	 login, dbcc_stmt.db_name)));
+		}
+	}
+
+	/* get physical schema name from logical schema name */
+	if (dbcc_stmt.schema_name)
+	{
+		schema_name = dbcc_stmt.schema_name;
+		nsp_name = get_physical_schema_name(db_name, dbcc_stmt.schema_name);
+	}
+	else
+	{
+		/* 
+		 * If schema_name is not provided, find default schema for current user
+		 * and get physical schema name
+		 */
+		guest_role_name = get_guest_role_name(db_name);
+		dbo_role_name = get_dbo_role_name(db_name);
+		
+		/* user will never be null here as cross-db calls are already handled */
+		Assert(user != NULL);
+
+		schema_name = get_authid_user_ext_schema_name((const char *) db_name, user);
```

Comment:
Please move it inside the third condition, It is not needed for every condition. Otherwise permissions related changes looks good.